system/lib/wasmfs/file.h

```diff
@@ -132,9 +132,8 @@ class DataFile : public File {
 
   // TODO: Allow backends to override the version of read with multiple iovecs
   // to make it possible to implement pipes. See #16269.
-  virtual __wasi_errno_t read(uint8_t* buf, size_t len, off_t offset) = 0;
-  virtual __wasi_errno_t
-  write(const uint8_t* buf, size_t len, off_t offset) = 0;
+  virtual ssize_t read(uint8_t* buf, size_t len, off_t offset) = 0;
+  virtual ssize_t write(const uint8_t* buf, size_t len, off_t offset) = 0;
```

Comment:
A comment on `ssize_t` here would be good. Is the idea that it is signed to allow either a size or a negative error code? (If so, how does tihs work on 32-bit, wouldn't it limit the size to 31 bits?

Reply:
Yes, these should return a non-negative accessed length or a negative error code, just like the `read` and `write` syscalls. I'll add a comment. The `ssize_t` does limit the access length to 31 bits, but that's also just like the system calls. Anyone reading more than two GB at a time can probably use a loop!

Reply:
Ah, sounds good, then it's ok if the syscalls have a limit here. I presume libc would make the second syscall for the user automatically.

Reply:
The libc functions _also_ return a ssize_t. Actually, the WASI syscalls we implement are the weird ones out of the bunch because they write the written bytes to an out param and return a positive error code.

Reply:
Oh, but isn't it possible to have files larger than `size_t`? Do libc methods just not allow reads and writes of that size despite the file being so big?

Reply:
Yes, files can be bigger than a `size_t`. You just can't read them all at once.